storage/src/main/java/org/apache/uniffle/storage/handler/impl/DataSkippableReadHandler.java

```diff
@@ -89,8 +89,10 @@ public ShuffleDataResult readShuffleData() {
       blocksOfSegment.and(expectBlockIds);
       if (!blocksOfSegment.isEmpty()) {
         // skip processed blockIds
-        blocksOfSegment.or(processBlockIds);
-        blocksOfSegment.xor(processBlockIds);
+        synchronized (processBlockIds) {
```

Comment:
This is a possible problem of introducing the lock, but I don't have a good idea to solve this

Reply:
Why do we need this if we don't use multiple threads?

Reply:
The separation of fetching(one thread)  and processing(another thread) will share the `processBlockIds.` 

For fetching, it will use the clientReadHandler to fetch data, but in its parent class, the `DataSkippableReadHandler` will skip some segments depending on the existence of `processBlockIds`.

And for processing, the `processBlockIds` indicates the block is  processed or not.

So, the race condition is not caused by the single thread but by the fetching and processing separated multiple threads.

Reply:
I  think it's unnecessary to use lock. In most cases, the blocks processed by the read and write threads will not be repeated.

Reply:
Emm... The key point is the `DataSkippableReadHandler` support duplicate blocks skip which will reduce the fetching time in some cases.

Reply:
> read and write threads

This means read and write processBlockIds.

I mean if the processing thread is processing block which id is `blockid1`, the probability that fetch thread reads `blockid1` again is very small at the same time.

Reply:
The race condition is caused on the `processBlockid` of no thread safe bitmap, rather than on single blockId.

Reply:
How about using `ReentrantReadWriteLock`?

Reply:
Yes, I also have similar thought, but I hope to avoid lock as much as possible, because in this case, the processing frequency is far greater than the fetching.

Reply:
@advancedxy WDYT?